Source/net_client.c

```diff
@@ -0,0 +1,1228 @@
+    SetRejectReason("Unknown reason");
+
+    while (client_connection.state == NET_CONN_STATE_CONNECTING)
+    {
+        int nowtime = I_GetTimeMS();
+
+        // Send a SYN packet every second.
+        if (nowtime - last_send_time > 1000 || last_send_time < 0)
+        {
+            NET_CL_SendSYN(data);
+            last_send_time = nowtime;
+        }
+
+        // time out after 5 seconds
+        if (nowtime - start_time > 5000)
+        {
+            SetRejectReason("No response from server");
+            break;
+        }
+
+        if (!sent_hole_punch && nowtime - start_time > 2000)
+        {
+            NET_Log("client: no response to SYN, requesting hole punch");
+            NET_RequestHolePunch(client_context, addr);
+            sent_hole_punch = true;
+        }
+
+        // run client code
+        NET_CL_Run();
+
+        // run the server, just in case we are doing a loopback connect
+        NET_SV_Run();
+
+        // Don't hog the CPU
+        I_Sleep(1);
+    }
+
+    if (client_connection.state == NET_CONN_STATE_CONNECTED)
+    {
+        // connected ok!
+        NET_Log("client: connected successfully");
+        SetRejectReason(NULL);
+        client_state = CLIENT_STATE_WAITING_LAUNCH;
+        drone = data->drone;
+
+        return true;
+    }
+    else
+    {
+        // failed to connect
+        NET_Log("client: failed to connect");
+        NET_CL_Shutdown();
+
+        return false;
+    }
+}
+
+// read game settings received from server
+
+boolean NET_CL_GetSettings(net_gamesettings_t *_settings)
+{
+    if (client_state != CLIENT_STATE_IN_GAME)
+    {
+        return false;
+    }
+
+    memcpy(_settings, &settings, sizeof(net_gamesettings_t));
+
+    return true;
+}
+
+// disconnect from the server
+
+void NET_CL_Disconnect(void)
+{
+    int start_time;
+
+    if (!net_client_connected)
+    {
+        return;
+    }
+
+    NET_Log("client: beginning disconnect");
+    NET_Conn_Disconnect(&client_connection);
+
+    start_time = I_GetTimeMS();
+
+    while (client_connection.state != NET_CONN_STATE_DISCONNECTED
+        && client_connection.state != NET_CONN_STATE_DISCONNECTED_SLEEP)
+    {
+        if (I_GetTimeMS() - start_time > 5000)
+        {
+            // time out after 5 seconds
+
+            NET_Log("client: no acknowledgement of disconnect received");
+            client_state = CLIENT_STATE_WAITING_START;
+
+            fprintf(stderr, "NET_CL_Disconnect: Timeout while disconnecting "
+                            "from server\n");
+            break;
+        }
+
+        NET_CL_Run();
+        NET_SV_Run();
+
+        I_Sleep(1);
+    }
+
+    // Finished sending disconnect packets, etc.
+    NET_Log("client: disconnect complete");
+    NET_CL_Shutdown();
+}
+
+void NET_CL_Init(void)
+{
+    // Try to set from the USER and USERNAME environment variables
+    // Otherwise, fallback to "Player"
+
+    if (net_player_name == NULL)
+    {
+        net_player_name = NET_GetRandomPetName();
```

Comment:
For some odd reason, I have my problems with this "pet name" thing. I guess it's too much of an Choco inside joke to apply here. Could you replace this initialization with a `GetUserName()` on WIN32 and `getlogin()` elsewhere?

Reply:
Maybe it's better to name them "Player 1" ... "Player 4"?

Reply:
Na, that's a bit lame, isn't it? 😉 I think even Vanilla had the sprite colors as player names, i.e. "green", "brown", "indigo", "red".

Reply:
Maybe we should invent dogs related joke here :smile:

Reply:
😮

Reply:
https://www.theodysseyonline.com/cliche-pet-names

Reply:
Even better:
https://www.dailypaws.com/dogs-puppies/dog-names/paw-patrol-dog-names

Reply:
Haha, that's nice.

Reply:
![net](https://user-images.githubusercontent.com/5077629/143603821-acd51642-b289-4572-91e3-fc33b0f3c70d.png)